src/xcode/ENA/ENA/Source/Services/Risk/Provider/Helper/DeviceTimeCheck.swift

```diff
@@ -0,0 +1,89 @@
+//
+// Corona-Warn-App
+//
+// SAP SE and all other contributors
+// copyright owners license this file to you under the Apache
+// License, Version 2.0 (the "License"); you may not use this
+// file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//
+
+import Foundation
+
+protocol DeviceTimeCheckProtocol {
+	func checkAndPersistDeviceTimeFlags(serverTime: Date, deviceTime: Date)
+	func resetDeviceTimeFlagsToDefault()
+}
+
+final class DeviceTimeCheck: DeviceTimeCheckProtocol {
+
+	// MARK: - Init
+
+	init(store: AppConfigCaching) {
+		self.store = store
+	}
+
+	// MARK: - Protocol DeviceTimeCheckProtocol
+
+	// MARK: - Internal
+
+	func checkAndPersistDeviceTimeFlags(serverTime: Date, deviceTime: Date) {
+		self.persistDeviceTimeCheckFlags(
+			deviceTimeIsCorrect: self.isDeviceTimeCorrect(
+				serverTime: serverTime,
+				deviceTime: deviceTime
+			),
+			deviceTimeCheckKillSwitchIsActive: self.isDeviceTimeCheckKillSwitchActive(
+				config: self.store.appConfig
+			)
+		)
+	}
+
+	func resetDeviceTimeFlagsToDefault() {
+		store.deviceTimeIsCorrect = true
+		store.deviceTimeErrorWasShown = false
+	}
+
+	// MARK: - Private
+
+	private let store: AppConfigCaching
+
+	private func persistDeviceTimeCheckFlags(
+		deviceTimeIsCorrect: Bool,
+		deviceTimeCheckKillSwitchIsActive: Bool
+	) {
+		store.deviceTimeIsCorrect = deviceTimeCheckKillSwitchIsActive ? true : deviceTimeIsCorrect
+		if store.deviceTimeIsCorrect {
+			store.deviceTimeErrorWasShown = false
+		}
+	}
+
+	private func isDeviceTimeCorrect(serverTime: Date, deviceTime: Date) -> Bool {
+		guard let serverTimeMinus2Hours = Calendar.current.date(byAdding: .hour, value: -2, to: serverTime),
+			  let serverTimePlus2Hours = Calendar.current.date(byAdding: .hour, value: 2, to: serverTime) else {
+			return true
```

Comment:
Does this consider that server time is in UTC?

Reply:
I think it should be fine. Date is a point in time. Adding 2 hours should result in a Date which is the same regardless of the timezone.
But to be 100% sure, I have changed the implementation to not use calendar. Instead use "addingTimeInterval" to manipulate the Date.